File: backend/app/schemas/inventory_verification.py
"""Inventory verification API schemas."""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import (
    INVENTORY_ITEM_TYPES,
    InventoryChangeRequestStatus,
    InventoryHistoryAction,
    InventoryItemType,
    InventoryVerificationStatus,
)


class InventoryItemsInput(BaseModel):
    model_config = ConfigDict(extra="forbid")

    shirts: int = Field(default=0, ge=0, le=500)
    trousers: int = Field(default=0, ge=0, le=500)
    sarees: int = Field(default=0, ge=0, le=500)
    jackets: int = Field(default=0, ge=0, le=500)
    bedsheets: int = Field(default=0, ge=0, le=500)
    blankets: int = Field(default=0, ge=0, le=500)
    curtains: int = Field(default=0, ge=0, le=500)
    other: int = Field(default=0, ge=0, le=500)


class InventoryItemLineResponse(BaseModel):
    item_type: InventoryItemType
    label: str
    quantity: int


class InventoryVerificationResponse(BaseModel):
    id: UUID
    order_id: UUID
    customer_id: UUID
    laundry_id: UUID
    status: InventoryVerificationStatus
    items: list[InventoryItemLineResponse]
    total_quantity: int
    recorded_by_user_id: UUID
    recorded_at: datetime
    confirmed_by_user_id: UUID | None
    confirmed_at: datetime | None
    locked_at: datetime | None
    is_locked: bool
    pending_change_request_id: UUID | None = None


class InventoryRecordRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    items: InventoryItemsInput
    note: str | None = Field(default=None, max_length=1000)


class InventoryChangeRequestInput(BaseModel):
    model_config = ConfigDict(extra="forbid")

    items: InventoryItemsInput
    reason: str = Field(min_length=10, max_length=2000)


class InventoryHistoryEntryResponse(BaseModel):
    id: UUID
    order_id: UUID
    verification_id: UUID
    action: InventoryHistoryAction
    items_snapshot: dict
    actor_user_id: UUID
    note: str | None
    created_at: datetime


class InventoryChangeRequestResponse(BaseModel):
    id: UUID
    order_id: UUID
    verification_id: UUID
    proposed_items: dict
    reason: str
    status: InventoryChangeRequestStatus
    requested_by_user_id: UUID
    reviewed_by_user_id: UUID | None
    reviewed_at: datetime | None
    admin_notes: str | None
    created_at: datetime


class InventoryChangeReviewRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    admin_notes: str | None = Field(default=None, max_length=2000)


ITEM_LABELS: dict[InventoryItemType, str] = {
    InventoryItemType.shirts: "Shirts",
    InventoryItemType.trousers: "Trousers",
    InventoryItemType.sarees: "Sarees",
    InventoryItemType.jackets: "Jackets",
    InventoryItemType.bedsheets: "Bedsheets",
    InventoryItemType.blankets: "Blankets",
    InventoryItemType.curtains: "Curtains",
    InventoryItemType.other: "Other Items",
}


def items_input_to_dict(items: InventoryItemsInput) -> dict[str, int]:
    return {t.value: int(getattr(items, t.value)) for t in INVENTORY_ITEM_TYPES}


def empty_items_dict() -> dict[str, int]:
    return {t.value: 0 for t in INVENTORY_ITEM_TYPES}

File: frontend/services/inventory-verification.ts
import { api, type ApiEnvelope } from '@/lib/api';
import type { DeliveryProofPhoto } from '@/services/delivery-proof';

export const INVENTORY_ITEM_TYPES = [
  'shirts',
  'trousers',
  'sarees',
  'jackets',
  'bedsheets',
  'blankets',
  'curtains',
  'other',
] as const;

export type InventoryItemType = (typeof INVENTORY_ITEM_TYPES)[number];

export const INVENTORY_ITEM_LABELS: Record<InventoryItemType, string> = {
  shirts: 'Shirts',
  trousers: 'Trousers',
  sarees: 'Sarees',
  jackets: 'Jackets',
  bedsheets: 'Bedsheets',
  blankets: 'Blankets',
  curtains: 'Curtains',
  other: 'Other Items',
};

export type InventoryItems = Record<InventoryItemType, number>;

export type InventoryVerificationStatus = 'pending_customer' | 'locked' | 'change_pending';

export interface InventoryItemLine {
  item_type: InventoryItemType;
  label: string;
  quantity: number;
}

export interface InventoryVerification {
  id: string;
  order_id: string;
  customer_id: string;
  laundry_id: string;
  status: InventoryVerificationStatus;
  items: InventoryItemLine[];
  total_quantity: number;
  recorded_by_user_id: string;
  recorded_at: string;
  confirmed_by_user_id: string | null;
  confirmed_at: string | null;
  locked_at: string | null;
  is_locked: boolean;
  pending_change_request_id: string | null;
}

export interface InventoryHistoryEntry {
  id: string;
  order_id: string;
  verification_id: string;
  action: string;
  items_snapshot: { items: InventoryItems; total: number };
  actor_user_id: string;
  note: string | null;
  created_at: string;
}

export interface InventoryChangeRequest {
  id: string;
  order_id: string;
  verification_id: string;
  proposed_items: { items: InventoryItems; total: number };
  reason: string;
  status: 'pending' | 'approved' | 'rejected';
  requested_by_user_id: string;
  reviewed_by_user_id: string | null;
  reviewed_at: string | null;
  admin_notes: string | null;
  created_at: string;
}

export function emptyInventoryItems(): InventoryItems {
  return {
    shirts: 0,
    trousers: 0,
    sarees: 0,
    jackets: 0,
    bedsheets: 0,
    blankets: 0,
    curtains: 0,
    other: 0,
  };
}

export function inventoryFromLines(lines: InventoryItemLine[]): InventoryItems {
  const base = emptyInventoryItems();
  for (const line of lines) {
    base[line.item_type] = line.quantity;
  }
  return base;
}

export function inventoryTotal(items: InventoryItems): number {
  return INVENTORY_ITEM_TYPES.reduce((sum, key) => sum + (items[key] ?? 0), 0);
}

export async function getPartnerInventoryVerification(orderId: string): Promise<InventoryVerification | null> {
  const { data } = await api.get<ApiEnvelope<InventoryVerification | null>>(
    `/partner/orders/${orderId}/inventory-verification`,
  );
  return data.data;
}

export async function recordPartnerInventory(
  orderId: string,
  items: InventoryItems,
  note?: string,
): Promise<InventoryVerification> {
  const { data } = await api.put<ApiEnvelope<InventoryVerification>>(
    `/partner/orders/${orderId}/inventory-verification`,
    { items, note },
  );
  return data.data;
}

export async function requestInventoryChange(
  orderId: string,
  items: InventoryItems,
  reason: string,
): Promise<InventoryChangeRequest> {
  const { data } = await api.post<ApiEnvelope<InventoryChangeRequest>>(
    `/partner/orders/${orderId}/inventory-verification/change-request`,
    { items, reason },
  );
  return data.data;
}

export async function getCustomerInventoryVerification(orderId: string): Promise<InventoryVerification | null> {
  const { data } = await api.get<ApiEnvelope<InventoryVerification | null>>(
    `/orders/${orderId}/inventory-verification`,
  );
  return data.data;
}

export async function confirmInventory(orderId: string): Promise<InventoryVerification> {
  const { data } = await api.post<ApiEnvelope<InventoryVerification>>(
    `/orders/${orderId}/inventory-verification/confirm`,
    {},
  );
  return data.data;
}

export async function listInventoryHistory(orderId: string): Promise<InventoryHistoryEntry[]> {
  const { data } = await api.get<ApiEnvelope<InventoryHistoryEntry[]>>(
    `/orders/${orderId}/inventory-verification/history`,
  );
  return data.data;
}

export async function listAdminInventoryChangeRequests(): Promise<InventoryChangeRequest[]> {
  const { data } = await api.get<ApiEnvelope<InventoryChangeRequest[]>>('/admin/inventory-change-requests');
  return data.data;
}

export async function approveInventoryChange(
  requestId: string,
  adminNotes?: string,
): Promise<InventoryVerification> {
  const { data } = await api.post<ApiEnvelope<InventoryVerification>>(
    `/admin/inventory-change-requests/${requestId}/approve`,
    { admin_notes: adminNotes },
  );
  return data.data;
}

export async function rejectInventoryChange(
  requestId: string,
  adminNotes?: string,
): Promise<InventoryChangeRequest> {
  const { data } = await api.post<ApiEnvelope<InventoryChangeRequest>>(
    `/admin/inventory-change-requests/${requestId}/reject`,
    { admin_notes: adminNotes },
  );
  return data.data;
}

import type { DisputeDetail, DisputeListItem } from '@/services/disputes';

export interface ComplaintListItem extends DisputeListItem {}
export interface ComplaintDetail extends DisputeDetail {}

export {
  listMyDisputes as listMyComplaints,
  getDisputeDetail as getComplaintDetail,
  fileDispute as createComplaint,
} from '@/services/disputes';

export type { DisputeListItem, DisputeDetail } from '@/services/disputes';
